### ISSUE TITLE
Crits does not start cleanly after being stopped via the docker stop command

### ISSUE THREAD

**Issue by jjs357** [REMnux/docker#7]
It looks like a docker start command after a Crits container was stopped fails because it tries to re-create the user nonroot even though this user already exists. If I look at the docker log for the container I see the following:
Traceback (most recent call last):
  File "/data/crits/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/data/crits/crits/core/management/commands/users.py", line 173, in handle
    user = CRITsUser.create_user(username, password, email)
  File "/data/crits/crits/core/user.py", line 679, in create_user
    user.create_api_key("default", analyst, default=True)
  File "/data/crits/crits/core/user.py", line 583, in create_api_key
    self.save(username=analyst)
  File "/data/crits/crits/core/crits_mongoengine.py", line 361, in _custom_save
    _refs=_refs)
  File "/usr/local/lib/python2.7/dist-packages/mongoengine/document.py", line 286, in save
    raise NotUniqueError(message % unicode(err))
mongoengine.errors.NotUniqueError: Tried to save duplicate unique keys (insertDocument :: caused by :: 11000 E11000 duplicate key error index: crits.users.$username_1  dup key: { : "nonroot" })

Can this be fixed easily? I am quite new to Docker and don't know where to start.

**Comment by wzod** [comment]
Hi James,

Can you provide further details on how you're using the Docker image?  Please detail the steps you performed that led to the error message.  Want to make sure I understand your use case/workflow.

Also, have you reviewed the documentation for the CRITs Docker image?

https://github.com/REMnux/docs/blob/master/tools/CRITs.md

**Comment by jjs357** [comment]
I had reviewed the documentation but didn’t fully appreciate the need to manually take the step of copying from /data/db to enable persistence.

I guess that is a way to realize some kind of data stability when the container is stopped and started.

What I have been doing with Laikaboss is to stop the container via docker and then start it again. Then attach to the named container.

This works, I am back where I left off.

If I do that with CRITS, I get an error — the one posted to the issue — and then the container shuts down on its own.

So I have to restart with the run command and then the password for nonroot is reset and the data in the CRITS db is gone. So I now see that the docs for the container say that I can copy the saved data from /data/db back to /data/db and I should then be able to see the content via the web app interface that was there prior to stopping the image. I will try this out real soon.

Can’t this behavior be automated in some way? I would need to keep the host defined workdir unchanged but I would expect that the container would be like I had left it.

I can use docker pause and unpause and then CRITS remains available when I attach. When I want to reboot the host, I need to stop rather than pause the container. But the docker attach mechanism does not work if I use docker stop and start. I want the CRITS container to act like a “normal” VM image as much as possible. I would like malware metadata to be retained across docker stops and starts within the container.

Thanks for your interest in this matter.

Jim

**Comment by jjs357** [comment]
I tried to copy the data. I get this:
nonroot@19c003af1ddc:~/workdir$ cp -r /data/db ~/workdir/
cp: cannot create regular file '/home/nonroot/workdir/db/mongod.lock': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/crits.0': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/local.0': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/local.ns': Permission denied
cp: cannot create directory '/home/nonroot/workdir/db/journal': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/crits.ns': Permission denied
cp: cannot create directory '/home/nonroot/workdir/db/_tmp': Permission denied

The directory db is created ok but none of the content is copied. Running docker on Mac OS X.

**Comment by jjs357** [comment]
I tried to build the docker container for CRITS myself from the latest docker file using docker on Mac OS X (Yosemite). I copied the dockerfile to a local directory. The build fails. I use the command:

docker build -t remnux/crits .

It gets to step 10 -- any help with this appreciated -- :

Step 10 : RUN bash -c 'mkdir -pv /data/{db,lock/apache2,log/apache2,log/supervisor,run/apache2,ssl/certs,ssl/private}' &&   cd /data/ &&   git clone https://github.com/crits/crits.git &&   cd crits/ &&   pip install -r requirements.txt
 ---> Running in 5763406d0028
mkdir: created directory '/data'
mkdir: created directory '/data/db'
mkdir: created directory '/data/lock'
mkdir: created directory '/data/lock/apache2'
mkdir: created directory '/data/log'
mkdir: created directory '/data/log/apache2'
mkdir: created directory '/data/log/supervisor'
mkdir: created directory '/data/run'
mkdir: created directory '/data/run/apache2'
mkdir: created directory '/data/ssl'
mkdir: created directory '/data/ssl/certs'
mkdir: created directory '/data/ssl/private'
Cloning into 'crits'...
Downloading/unpacking Django<1.7 (from -r requirements.txt (line 1))
Downloading/unpacking Pillow (from -r requirements.txt (line 2))
  Running setup.py (path:/tmp/pip_build_root/Pillow/setup.py) egg_info for package Pillow
    Single threaded build, not installing mp_compile:1 processes
    
Downloading/unpacking amqp (from -r requirements.txt (line 3))
  Running setup.py (path:/tmp/pip_build_root/amqp/setup.py) egg_info for package amqp
    
Downloading/unpacking anyjson (from -r requirements.txt (line 4))
  Downloading anyjson-0.3.3.tar.gz
  Running setup.py (path:/tmp/pip_build_root/anyjson/setup.py) egg_info for package anyjson
    
Downloading/unpacking billiard (from -r requirements.txt (line 5))
  Running setup.py (path:/tmp/pip_build_root/billiard/setup.py) egg_info for package billiard
    
    warning: no files found matching '*.py' under directory 'Lib'
Downloading/unpacking biplist (from -r requirements.txt (line 6))
  Downloading biplist-0.9.tar.gz
  Running setup.py (path:/tmp/pip_build_root/biplist/setup.py) egg_info for package biplist
    
Downloading/unpacking celery (from -r requirements.txt (line 7))
Downloading/unpacking defusedxml (from -r requirements.txt (line 8))
  Running setup.py (path:/tmp/pip_build_root/defusedxml/setup.py) egg_info for package defusedxml
    
Downloading/unpacking django-celery (from -r requirements.txt (line 9))
  Running setup.py (path:/tmp/pip_build_root/django-celery/setup.py) egg_info for package django-celery
    
    warning: no files found matching '*' under directory 'bin'
    no previously-included directories found matching 'bin/*.pyc'
    no previously-included directories found matching 'tests/*.pyc'
    no previously-included directories found matching 'docs/*.pyc'
    no previously-included directories found matching 'extra/*.pyc'
    no previously-included directories found matching 'djcelery/*.pyc'
    no previously-included directories found matching 'docs/.build'
    no previously-included directories found matching 'examples/*.pyc'
Downloading/unpacking django-tastypie==0.11.0 (from -r requirements.txt (line 10))
Downloading/unpacking django-tastypie-mongoengine==0.4.5 (from -r requirements.txt (line 11))
  Running setup.py (path:/tmp/pip_build_root/django-tastypie-mongoengine/setup.py) egg_info for package django-tastypie-mongoengine
    
    warning: no files found matching '*.txt'
Downloading/unpacking kombu (from -r requirements.txt (line 12))
Downloading/unpacking lxml (from -r requirements.txt (line 13))
  Running setup.py (path:/tmp/pip_build_root/lxml/setup.py) egg_info for package lxml
    /usr/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'bugtrack_url'
      warnings.warn(msg)
    Building lxml version 3.4.4.
    Building without Cython.
    Using build configuration of libxslt 1.1.28
    
    warning: no previously-included files found matching '*.py'
Downloading/unpacking m2crypto (from -r requirements.txt (line 14))
  Running setup.py (path:/tmp/pip_build_root/m2crypto/setup.py) egg_info for package m2crypto
    
    warning: no previously-included files found matching 'M2Crypto/_m2crypto.py'
Downloading/unpacking mongoengine==0.8.7 (from -r requirements.txt (line 16))
  Running setup.py (path:/tmp/pip_build_root/mongoengine/setup.py) egg_info for package mongoengine
    0.8.7
    
    no previously-included directories found matching 'docs/_build'
Downloading/unpacking pydeep==0.2 (from -r requirements.txt (line 17))
  Downloading pydeep-0.2.tar.gz
  Running setup.py (path:/tmp/pip_build_root/pydeep/setup.py) egg_info for package pydeep
    
Downloading/unpacking pymongo==2.7.2 (from -r requirements.txt (line 18))
  Running setup.py (path:/tmp/pip_build_root/pymongo/setup.py) egg_info for package pymongo
    
Downloading/unpacking pyparsing (from -r requirements.txt (line 19))
  Downloading pyparsing-2.0.5-py2.py3-none-any.whl
Downloading/unpacking python-dateutil (from -r requirements.txt (line 20))
Downloading/unpacking python-ldap (from -r requirements.txt (line 21))
  Running setup.py (path:/tmp/pip_build_root/python-ldap/setup.py) egg_info for package python-ldap
    defines: HAVE_SASL HAVE_TLS HAVE_LIBLDAP_R
    extra_compile_args:
    extra_objects:
    include_dirs: /usr/include /usr/include/sasl /usr/local/include /usr/local/include/sasl
    library_dirs: /usr/lib /usr/lib64 /usr/local/lib /usr/local/lib64
    libs: ldap_r
    
    file Lib/ldap.py (for module ldap) not found
    file Lib/ldap/controls.py (for module ldap.controls) not found
    file Lib/ldap/extop.py (for module ldap.extop) not found
    file Lib/ldap/schema.py (for module ldap.schema) not found
    warning: no files found matching 'Makefile'
    warning: no files found matching 'Modules/LICENSE'
Downloading/unpacking python-magic==0.4.6 (from -r requirements.txt (line 22))
  Downloading python-magic-0.4.6.tar.gz
  Running setup.py (path:/tmp/pip_build_root/python-magic/setup.py) egg_info for package python-magic
    
Downloading/unpacking python-mimeparse (from -r requirements.txt (line 23))
  Downloading python-mimeparse-0.1.4.tar.gz
  Running setup.py (path:/tmp/pip_build_root/python-mimeparse/setup.py) egg_info for package python-mimeparse
    
Downloading/unpacking pytz (from -r requirements.txt (line 24))
Downloading/unpacking pyyaml (from -r requirements.txt (line 25))
  Running setup.py (path:/tmp/pip_build_root/pyyaml/setup.py) egg_info for package pyyaml
    
Requirement already satisfied (use --upgrade to upgrade): requests in /usr/lib/python2.7/dist-packages (from -r requirements.txt (line 26))
Requirement already satisfied (use --upgrade to upgrade): setuptools in /usr/lib/python2.7/dist-packages (from -r requirements.txt (line 27))
Downloading/unpacking simplejson (from -r requirements.txt (line 28))
  Running setup.py (path:/tmp/pip_build_root/simplejson/setup.py) egg_info for package simplejson
    
Requirement already satisfied (use --upgrade to upgrade): six in /usr/lib/python2.7/dist-packages (from -r requirements.txt (line 29))
Downloading/unpacking ushlex (from -r requirements.txt (line 30))
  Downloading ushlex-0.99.tar.gz
  Running setup.py (path:/tmp/pip_build_root/ushlex/setup.py) egg_info for package ushlex
    
Requirement already satisfied (use --upgrade to upgrade): wsgiref in /usr/lib/python2.7 (from -r requirements.txt (line 31))
Downloading/unpacking olefile (from -r requirements.txt (line 32))
  Running setup.py (path:/tmp/pip_build_root/olefile/setup.py) egg_info for package olefile
    
Installing collected packages: Django, Pillow, amqp, anyjson, billiard, biplist, celery, defusedxml, django-celery, django-tastypie, django-tastypie-mongoengine, kombu, lxml, m2crypto, mongoengine, pydeep, pymongo, pyparsing, python-dateutil, python-ldap, python-magic, python-mimeparse, pytz, pyyaml, simplejson, ushlex, olefile
  Running setup.py install for Pillow
    Single threaded build, not installing mp_compile:1 processes
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip_build_root/Pillow/setup.py", line 767, in <module>
        zip_safe=not debug_build(),
      File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/dist-packages/setuptools/command/install.py", line 53, in run
        return _install.run(self)
      File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run
        self.run_command('build')
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
        self.run_command(cmd_name)
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build_ext.py", line 337, in run
        self.build_extensions()
      File "/tmp/pip_build_root/Pillow/setup.py", line 515, in build_extensions
        % (f, f))
    ValueError: --enable-jpeg requested but jpeg not found, aborting.
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/Pillow/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-Vh7KN7-record/install-record.txt --single-version-externally-managed --compile:
    Single threaded build, not installing mp_compile:1 processes

running install

running build

running build_py

creating build

creating build/lib.linux-x86_64-2.7

creating build/lib.linux-x86_64-2.7/PIL

copying PIL/WalImageFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/EpsImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/FpxImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageChops.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/Hdf5StubImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/CurImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PcdImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/OleFileIO.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/features.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/SgiImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageGrab.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/TgaImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/XpmImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageSequence.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/JpegPresets.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/McIdasImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageEnhance.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImagePath.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageDraw.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PaletteFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/IcoImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PcxImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageColor.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/XbmImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/GimpPaletteFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/FontFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageQt.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/BmpImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/GribStubImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/TarIO.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/_binary.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageTransform.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/MspImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PyAccess.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/FliImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ContainerIO.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ExifTags.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/XVThumbImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PpmImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/MpoImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/GbrImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PcfFontFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageDraw2.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageFilter.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/_util.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/MpegImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/SpiderImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/WebPImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PdfImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/JpegImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/GifImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/FitsStubImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/IcnsImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PalmImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/Jpeg2KImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageWin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/BufrStubImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageStat.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImtImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PsdImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PSDraw.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/WmfImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImagePalette.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/SunImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/DcxImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/MicImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PngImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageMath.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/IptcImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageTk.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageMorph.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageCms.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageOps.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/__init__.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/GdImageFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageMode.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/PixarImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageShow.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/TiffTags.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/BdfFontFile.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/Image.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/ImageFont.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/TiffImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL

copying PIL/GimpGradientFile.py -> build/lib.linux-x86_64-2.7/PIL

running egg_info

writing Pillow.egg-info/PKG-INFO

writing top-level names to Pillow.egg-info/top_level.txt

writing dependency_links to Pillow.egg-info/dependency_links.txt

warning: manifest_maker: standard file '-c' not found



reading manifest file 'Pillow.egg-info/SOURCES.txt'

reading manifest template 'MANIFEST.in'

writing manifest file 'Pillow.egg-info/SOURCES.txt'

copying PIL/OleFileIO-README.md -> build/lib.linux-x86_64-2.7/PIL

running build_ext

Traceback (most recent call last):

  File "<string>", line 1, in <module>

  File "/tmp/pip_build_root/Pillow/setup.py", line 767, in <module>

    zip_safe=not debug_build(),

  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup

    dist.run_commands()

  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands

    self.run_command(cmd)

  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command

    cmd_obj.run()

  File "/usr/lib/python2.7/dist-packages/setuptools/command/install.py", line 53, in run

    return _install.run(self)

  File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run

    self.run_command('build')

  File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command

    self.distribution.run_command(command)

  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command

    cmd_obj.run()

  File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run

    self.run_command(cmd_name)

  File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command

    self.distribution.run_command(command)

  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command

    cmd_obj.run()

  File "/usr/lib/python2.7/distutils/command/build_ext.py", line 337, in run

    self.build_extensions()

  File "/tmp/pip_build_root/Pillow/setup.py", line 515, in build_extensions

    % (f, f))

ValueError: --enable-jpeg requested but jpeg not found, aborting.

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/Pillow/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-Vh7KN7-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip_build_root/Pillow
Storing debug log for failure in /root/.pip/pip.log
The command '/bin/sh -c bash -c 'mkdir -pv /data/{db,lock/apache2,log/apache2,log/supervisor,run/apache2,ssl/certs,ssl/private}' &&   cd /data/ &&   git clone https://github.com/crits/crits.git &&   cd crits/ &&   pip install -r requirements.txt' returned a non-zero code: 1

**Comment by jjs357** [comment]
I have added some more comments to the docs site. Could not build CRITS from a local copy of the dockerfile. I must be doing something wrong.

Jim

**Comment by wzod** [comment]
HI Jim,

Regarding the permission denied errors when copying the data/db, did you make the ~/crits-data directory world-accessible (e.g. chmod a+xwr ~/crits-data) prior to running  the Docker image?

Regarding issues building the Docker image, I recently (yesterday) ran into similar issues when building the Docker image; it appears there are some dependencies that are now required for the build to complete (there were no issues when we built the Docker image back in early October). I'm currently working on a fix to resolve the issues with building the Docker image and should have that available soon.

In the meantime, please use the Docker image from the REMnux repo as that build should work.  Considering your use case, you should be able to copy the data while using the Docker image to your workdir directory and use the ~/crits-data directory, which contains your changes, when returning to the Docker image.

Please let me know if making the ~/crits-data directory world-accessible resolves the permissions issue.

Hope this helps!

**Comment by jjs357** [comment]
Yes, my permissions are set correctly for crits-data from the host side. But inside the container, these permissions are not honored. I can't seem to make the changes to the nonroot user (like changing UID to 1000) to get me past the error. The copies seem to made as the user 1000 with group staff. Hence the copies fail. This could be a Mac OS X docker glitch.

The REMnux repo docker image itself works but the copying steps fail for me. So I am stuck for now.

**Comment by jjs357** [comment]
I just tried this all again. Made sure that the crits-data (bound to workdir) permissions are world writeable.

There are other docker container (not CRITS) persistence problems that are solved by changing the UID of the user  running supervisord to 1000. I see in the startup messages for the CRITS container that the there is a set UID command setting to the value 999 which is the default UID of nonroot. The problem is that the workdir permissions look like:

nonroot@2934ee3c22ee:~$ ls -lt
total 0
drwxrwxrwx 1 1000 staff 136 Nov  2 21:07 workdir

When I run the copy command, I do get a new subdirectory db but its permissions are:

nonroot@2934ee3c22ee:~$ ls -l workdir/
total 8
drwxr-xr-x 1 1000 staff   68 Nov  4 17:02 db

So since the copy is running as user nonroot, the copies fail:

nonroot@2934ee3c22ee:~$ cp -r /data/db ~/workdir/
cp: cannot create regular file '/home/nonroot/workdir/db/mongod.lock': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/crits.0': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/local.0': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/local.ns': Permission denied
cp: cannot create directory '/home/nonroot/workdir/db/journal': Permission denied
cp: cannot create regular file '/home/nonroot/workdir/db/crits.ns': Permission denied
cp: cannot create directory '/home/nonroot/workdir/db/_tmp': Permission denied

Since nonroot is not in the sudo'rs definition, I can't try things as root.

**Comment by jjs357** [comment]
I found a work-around to my permissions problem. There is a "shim" VM that is part of the docker toolbox for Mac OS X and this VM has some configuration options for the uid and gid used when accessing the /Users directory under Mac OS. Since my workdir is in /Users/jjs, I was able to add a configuration option in this VM that sets the uid and gid for all of /Users to 999:999. Then when I run the CRITS container, the specific directory that I am sharing as crits-data is shown as being owned by nonroot:nonroot. Then these cp commands do work. The CRITS container somehow inherits the uid and gid settings.

At least I can get some measure of persistence between restarts of the container.

Still would be nice if I could stop and start the docker container and not getting an error when doing so.
